source/s3_meta_request.c

```diff
@@ -1781,7 +1843,21 @@ static void s_s3_meta_request_event_delivery_task(struct aws_task *task, void *a
                     if (meta_request->meta_request_level_running_response_sum) {
                         aws_checksum_update(meta_request->meta_request_level_running_response_sum, &response_body);
                     }
-                    if (meta_request->body_callback != NULL &&
+                    if (meta_request->recv_file) {
+                        /* Write the data directly to the file. No need to seek, since the event will always be
+                         * delivered with the right order. */
+                        if (fwrite((void *)response_body.ptr, response_body.len, 1, meta_request->recv_file) < 1) {
+                            int errno_value = ferror(meta_request->recv_file) ? errno : 0; /* Always cache errno  */
+                            error_code = aws_translate_and_raise_io_error_or(errno_value, AWS_ERROR_FILE_WRITE_FAILURE);
```

Comment:
aws_translate_and_raise_io_error_or() is like aws_raise_error(): it always returns AWS_OP_ERR(-1)
```suggestion
                            aws_translate_and_raise_io_error_or(errno_value, AWS_ERROR_FILE_WRITE_FAILURE);
                            error_code = aws_last_error()
```